README.md
# OpenGTS
Fork of OpenGTS tracking system.
A new livetracking system for free flying will be developed.
The system will be compatible with the original OpenGTS to facilitate future updates from mainstream.
Maybe some features will be pushed to upstream.


OpenGTS/CHANGELOG.txt
-----------------------------------------------------------------------------------
GTS Changelog
Copyright 2007-2014 GeoTelematic Solutions, Inc.
-----------------------------------------------------------------------------------

Version 2.5.7 [2014/10/07]
- NEW: Added additional status codes.
- NEW: Initial support for DeviceGroup notify email address (incomplete).
- NEW: Added "useResultAddress" property to "nominatim" reverse-geocode provider.
- NEW: Added DATA_VEHICLE_MAKE, DATA_VEHICLE_MODEL for report FieldLayout.
- NEW: Added "fleetSortByField" property to MotionReport.
- FIX: Fixed odometer offset used by MotionReport.
- FIX: Fixed ignition-hour accoumulation in "Device.java".
- UPD: Check for "Timezones" tag within "Domain" in private.xml
- UPD: Changes made to support compiling under Java-8.
- UPD: Added comments to "org/opengts/war/tools/BufferedHttpServletResponse.java"
       to indicate the change required in this module to support compiling with
       Tomcat-8 (changes required due to API differences in Servlet Spec 3.1).
- DCS: tk10x] Added support for using the last valid GPS location.
- DCS: tk10x] Added support for VJoy device.
- DCS: tk10x] Cell-tower data parsing for tk103_2 (thanks to Franjieh El Khoury).

Version 2.5.6 [2014/06/29]
- NEW: Enabled "Driver Message Detail" report to display messages returned from
       from CelltracGTS/Pro application.
- NEW: Added EventData report columns "ptoDistance", "workHours", "workDistance".
- NEW: Added EventData columns FLD_ptoDistanceKM, FLD_workDistanceKM
- NEW: Added initial support for FuelLevelProfile.
- NEW: Initial implementation for active/inactive Geozone checks.
- FIX: Fixed battery-level percentage display on some reports.
- FIX: Fixed NPE that would occasionally occur when displaying "Near By" Geozones.

Version 2.5.5 [2014/05/05]
- NEW: Added additional status codes.
- NEW: Added new fields to EventData table
- NEW: Added support for fuel-cost to Account/Device table and reports.
- DCS: gprmc] Added battery-temperature support.
- DCS: gprmc] Automatically adjust non-decimal battery level value to a percent.

Version 2.5.4 [2014/04/02]
- NEW: Added additional status codes.
- NEW: Added "allowWebService" column to Account table.
- NEW: Added "insuranceExpire" column to Device table.
- UPD: Additional changes to support table management of indexes in MS SQL Server.
- UPD: Update CheckInstall to list currently loaded config files.
- DCS: gc101] Fix packets sent via POST
- DCS: gprmc] Fix packets sent via POST
- DCS: sanav] Added support for checking "maximumHDOP".
- DCS: tk10x] Added support for new TK102 protocol (TK102B).

Version 2.5.3 [2014/01/01]
- NEW: Added additional status codes.
- NEW: Added generalized replacement handling to SMSOutboundGateway.java
- NEW: "Driving/Stopped Time Summary" report supported in Fleet/Group detail reports.
- NEW: Added disk utilization checking to "checkInstall" command.
- NEW: Added "equipmentStatus" / "licenseExpire" columns to Device table.
- NEW: Added additional fields to report FieldLayout.
- NEW: Added fleet detail reporting support to MotionReport.
- CHG: "Driving/Stopped Time Summary" report moved to Device/Vehicle Detail reports.
- DCS: gprmc] Added support for CellTower lookup (thanks to STefan Mayer)
- DCS: template] fixed "createGPSEvent" to property set "lastModemID".

Version 2.5.2 [2013/09/30]
- CHG: Sample data ("demo" account) moved to 2013/08/26
- NEW: Added additional status codes.
- NEW: Added support for updating key fields.
- UPD: Increased PendingCommands "commandArgs" field length.
- UPD: Updated "checkInstall" MySQL check to also look for "/etc/mysql/my.cnf"
- UPD: Added field length check to EventData "postalCode".
- UPD: Added "DATA_DRIVER_ID" field option to TripReport.
- DCS: tk10x] added initial support for "TK-nano" device (ASCII packets).
- DCS: gprmc] fixed date/time blank issue.

Version 2.5.1 [2013/08/06]
- NEW: Added additional status codes.
- NEW: CheckInstall] Group displayed reports by type.
- NEW: Added Account "allowNotify" field.
- NEW: Added Account command-line parameter "-setPassword".
- NEW: Added Device "lastEngineOnTime"/"lastEngineOffTime" fields.
- NEW: Added Device engine-hour accumulation based on Engine On/Off events.
- NEW: Added Device-based work-hour reporting to Motion Report.
- NEW: Added unique SIM-phone-number support to Device/Vehicle Admin.
- FIX: Fixed Motion Report idle time stop calculation.
- NEW: Added additional report column support to "FieldLayout.java"
- FIX: Fix NPE in "DateTime.parseDateTime", which could occur under some conditions.
- FIX: Fixe "Payload.readLongBCD" to support reading Little-Endian values.
- UPD: Added ACL support for Read/View to "System Accounts".
- DCS: gprmc] Added battery-level/volts, #satellites (thanks to STefan Mayer)

Version 2.5.0 [2013/06/05]
- NEW: Added support for obtaining an approximate count of table records when using
       MySQL "InnoDB" ("InnoDB" does not efficiently support the "count(*)" select
       clause for obtaining the exact number of table records).
- NEW: Added support for display MySQL engine type to "CheckInstall.java"
- NEW: Added DateTime support for parsing delta-month/year dates (ie. "+1m", "-2y")
- NEW: Added additional status codes.
- DCS: Fixed GPSEvent status code logging.
- DCS: template] Fixed "$POS,..." longitude parsing.
- DCS: template] Fixed possible NPE occurance in "createGPSEvent"
- DCS: tk10x] Fixed mileage/odometer calculation.
- DCS: tk10x] Added handling of GPIO events to "parseInsertRecord_TK103_alt"

Version 2.4.9 [2013/04/08]
- NEW: Added additional status codes.
- FIX: Fixed JSON recursive array parsing.
- NEW: Added RTProperties optional property setting feature (ie. "key?=value")
- NEW: Added additional fields to reporting EventDataLayout.

Version 2.4.8 [2013/03/01]
- NEW: Added additional status codes.
- NEW: Ignore old events when processing event Geozone detection.

OpenGTS/src/org/opengts/Version.java
// ----------------------------------------------------------------------------
// Copyright 2007-2014, GeoTelematic Solutions, Inc.
// All rights reserved
// ----------------------------------------------------------------------------
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
// 
// http://www.apache.org/licenses/LICENSE-2.0
// 
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// ----------------------------------------------------------------------------
// Change History:
//  2007/01/25 v1.2.0  [Initial release]
//  2007/02/08 v1.2.1,  2007/02/11 v1.3.0,  2007/02/16 v1.3.1,  2007/02/18 v1.3.2
//  2007/02/26 v1.3.3,  2007/02/28 v1.3.4,  2007/03/11 v1.3.5,  2007/03/16 v1.3.6,
//  2007/03/25 v1.3.7,  2007/03/30 v1.4.0,  2007/04/15 v1.4.1,  2007/04/17 v1.4.2,
//  2007/05/06 v1.4.3,  2007/05/20 v1.4.4,  2007/05/25 v1.4.5,  2007/06/03 v1.4.6,
//  2007/06/13 v1.4.7,  2007/06/14 v1.4.8,  2007/06/30 v1.5.0,  2007/07/14 v1.5.1,
//  2007/07/27 v1.5.2,  2007/08/09 v1.5.3,  2007/09/16 v1.6.0,  2007/11/28 v1.6.1,
//  2007/12/13 v1.6.2,  2008/01/10 v1.6.3,  2008/02/04 v1.6.4,  2008/02/07 v1.6.5,
//  2008/02/11 v1.6.6,  2008/02/17 v1.6.7,  2008/02/21 v1.6.8,  2008/02/27 v1.6.9,
//  2008/03/12 v1.6.10, 2008/03/28 v1.7.0,  2008/04/11 v1.7.1,  2008/05/14 v1.8.0,
//  2008/05/20 v1.8.1,  2008/05/22 v1.8.2,  2008/06/20 v1.8.3,  2008/07/08 v1.8.4,
//  2008/07/21 v1.9.0,  2008/07/27 v1.9.1,  2008/08/08 v1.9.2,  2008/08/15 v1.9.3,
//  2008/08/17 v1.9.4,  2008/08/20 v1.9.5,  2008/08/24 v1.9.6,  2008/09/01 v1.9.7,
//  2008/09/12 v1.9.8,  2008/09/19 v2.0.1,  2008/10/16 v2.0.2,  2008/12/01 v2.0.3,
//  2009/01/01 v2.0.4,  2009/02/01 v2.0.5,  2009/02/20 v2.0.6,  2009/04/02 v2.0.7,
//  2009/05/01 v2.0.8,  2009/05/24 v2.0.9,  2009/05/27 v2.1.0,  2009/06/01 v2.1.1,
//  2009/07/01 v2.1.2,  2009/08/09 v2.1.3,  2009/08/23 v2.1.4,  2009/08/25 v2.1.4.1,
//  2009/09/23 v2.1.5,  2009/10/05 v2.1.6,  2009/11/01 v2.1.7,  2009/11/10 v2.1.8,
//  2009/12/16 v2.1.9,  2010/01/29 v2.2.0,  2010/04/11 v2.2.1,  2010/04/25 v2.2.2,
//  2010/05/24 v2.2.3,  2010/06/17 v2.2.4,  2010/07/04 v2.2.5,  2010/07/18 v2.2.6,
//  2010/09/10 v2.2.7,  2010/10/25 v2.2.8,  2010/11/29 v2.2.9,  2011/01/28 v2.3.0,
//  2011/03/08 v2.3.1,  2011/04/01 v2.3.2,  2011/05/15 v2.3.3,  2011/06/16 v2.3.4,
//  2011/07/01 v2.3.5,  2011/07/15 v2.3.6,  2011/08/21 v2.3.7,  2011/10/10 v2.3.8,
//  2011/12/06 v2.3.9,  2012/02/03 v2.4.0,  2012/04/20 v2.4.1,  2012/05/27 v2.4.2,
//  2012/06/29 v2.4.3,  2012/08/01 v2.4.4,  2012/09/02 v2.4.5,  2012/11/02 v2.4.6,
//  2012/12/27 v2.4.7,  2013/03/01 v2.4.8,  2013/04/08 v2.4.9,  2013/06/05 v2.5.0,
//  2013/08/06 v2.5.1,  2013/09/30 v2.5.2,  2014/01/01 v2.5.3,  2014/03/03 v2.5.4,
//  2014/05/05 v2.5.5,  2014/06/29 v2.5.6,
//  2014/10/07 v2.5.7
// ----------------------------------------------------------------------------
package org.opengts;

import java.util.Date;
import java.util.Locale;
import java.util.TimeZone;
import java.text.SimpleDateFormat;

import org.opengts.util.*;

/**
*** This class provides runtime version information to the OpenGTS modules.
**/

public class Version
{

    // ------------------------------------------------------------------------

    // "public" as of 2.3.4-B06
    public  static final String COPYRIGHT = "Copyright 2007-2014, GeoTelematic Solutions, Inc.";

    // ------------------------------------------------------------------------

    // This string is parsed via 'grep' & 'sed' scripts.
    // ONLY the version value specified within the quotes should change.
    private static final String VERSION = "2.5.7"; // 2.5.7-B45"; // 2.5.6-B27

    // This public constant should only be accessed externally by the 'GTSAdmin' application.
    public  static final String GTS_ENTERPRISE_PREFIX = "E";
    public  static final String COMPILED_VERSION = 
        ////  GTS_ENTERPRISE_PREFIX + // "Enterprise"
        VERSION;

    // last compile time
    private static final long COMPILE_TIMESTAMP = CompileTime.COMPILE_TIMESTAMP;

    // ------------------------------------------------------------------------

    // package release (modified by command-line 'sed' script to insert actual epoch time)
    private static final long PACKAGE_TIMESTAMP = 1412800495L;
    public static final String SERVICE_ACCOUNT_ID = "opengts";

    // ------------------------------------------------------------------------
    // ------------------------------------------------------------------------

    /**
    *** Initializes the RTConfig constant 'version' property.
    **/
    public static void initVersionProperty()
    {
        RTProperties constantProps = RTConfig.getRuntimeConstantProperties();
        if (!constantProps.hasProperty(RTKey.VERSION)) {
            constantProps.setProperty(RTKey.VERSION, Version.getVersion());
            //System.out.println("Set Version = " + Version.getVersion());
        } else {
            // already initialized, no need to reinitialize
        }
    };

    // ------------------------------------------------------------------------
    // ------------------------------------------------------------------------

    /**
    *** Returns the copyright notice
    *** @return The copyright notice
    **/
    public static String getCopyright()
    {
        return COPYRIGHT;
    }

    /** 
    *** Returns the compiled version String
    *** @return The version String
    **/
    public static String getVersion()
    {
        return COMPILED_VERSION;
    }

    /** 
    *** Returns the compiled package release timestamp
    *** @return The package release timestamp
    **/
    public static long getPackageTimestamp()
    {
        return PACKAGE_TIMESTAMP;
    }

    /** 
    *** Returns the last compiled timestamp
    *** @return The last compiled timestamp
    **/
    public static long getCompileTimestamp()
    {
        return COMPILE_TIMESTAMP;
    }

    /**
    *** Returns the last compiled time date string
    *** @return The last compiled time date string
    **/
    public static String getCompileTime()
    {
        // (new DateTime(CompileTime.COMPILE_TIMESTAMP))
        return (new DateTime(COMPILE_TIMESTAMP)).toString();
    }

    /** 
    *** Returns the ServiceAccountID
    *** @return The ServiceAccountID
    **/
    public static String getServiceAccountID()
    {
        String sai = RTConfig.getString("ServiceAccount.ID","?");
        return sai.equals(SERVICE_ACCOUNT_ID)? sai : (sai+"("+SERVICE_ACCOUNT_ID+")");
    }

    // ------------------------------------------------------------------------

    /**
    *** Returns a String representation of this Version
    *** @return A String representation of this Version
    **/
    public static String getInfo()
    {
        StringBuffer sb = new StringBuffer();
        sb.append(Version.COPYRIGHT).append("\n");
        sb.append("Version: ").append(Version.getVersion()).append(" [").append(getServiceAccountID()).append("]").append("\n");
        SimpleDateFormat dateFmt = new SimpleDateFormat("yyyy/MM/dd HH:mm:ss zzz", Locale.US);
        if (PACKAGE_TIMESTAMP > 0L) {
            String dfmt = dateFmt.format(new Date(PACKAGE_TIMESTAMP * 1000L));
            sb.append("Package: [" + PACKAGE_TIMESTAMP + "] " + dfmt + "\n");
        }
        if (COMPILE_TIMESTAMP > 0L) {
            String dfmt = dateFmt.format(new Date(COMPILE_TIMESTAMP * 1000L));
            sb.append("Compile: [" + COMPILE_TIMESTAMP + "] " + dfmt + "\n");
        }
        return sb.toString();
    }

    // ------------------------------------------------------------------------

    /**
    *** Main entry point to display the current version
    *** @param argv The command-line args
    **/
    public static void main(String argv[])
    {
        // do not load RTConfig here
        if ((argv.length > 0) && argv[0].equals("-info")) {
            // Version: 2.0.5
            // Compile: [1211915119] 2008/05/27 19:36:38 GMT
            // Package: [1211915119] 2008/05/27 19:36:38 GMT
            System.out.println(Version.getInfo());
        } else
        if ((argv.length > 0) && argv[0].equals("-package")) {
            // "1211915119"
            System.out.println(Version.getPackageTimestamp());
        } else
        if ((argv.length > 0) && argv[0].equals("-compile")) {
            // "1211915119"
            System.out.println(Version.getCompileTimestamp());
        } else {
            // "1.8.3"
            System.out.println(Version.getVersion());
        }
    }

    // ------------------------------------------------------------------------

}
